### PR TITLE
OLS-363: Provider parameters check

### PR DESCRIPTION
**Pull request by tisnik**
## Description

Provider parameters check

## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [ ] Documentation Update
- [x] Configuration Update
- [ ] Bump-up dependent library

## Related Tickets & Documents

- Related Issue #[OLS-363](https://issues.redhat.com//browse/OLS-363)

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] PR has passed all pre-merge test jobs.
- [ ] If it is a core feature, I have added thorough tests.

### CODE REVIEW

File: ols/src/llms/providers/provider.py [hunk 1]
```diff
@@ -6,12 +6,34 @@
...
+from ols.app.models.config import InvalidConfigurationError, ProviderConfig
+from ols.constants import (
+    PROVIDER_AZURE_OPENAI,
+)
 from ols.utils import config
 
 logger = logging.getLogger(__name__)
 
 
+class ProviderParameter:
```
Comment (onmete) [inline ols/src/llms/providers/provider.py:18]: What about pydantic class or dataclass?

File: ols/src/llms/providers/provider.py [hunk 2]
```diff
@@ -44,6 +66,33 @@ def __init__(
         self.model = model
         self.provider_config = provider_config
         self.params = self._override_params(params or {})
+        # self._validate_parameters()
+
+    def _validate_parameters(self) -> None:
+        """Validate parameters for LLM provider."""
+        provider = self.provider_config.type
+
+        if provider is None:
```
Comment (onmete) [inline ols/src/llms/providers/provider.py:75]: Can this happen now? If not specified in the config, the name of the provider is used as type.

Comment (onmete) [reply ols/src/llms/providers/provider.py:75]: Scratch that. The name can be unset -> None. It is weird but our config allows that...

File: ols/src/llms/providers/provider.py [hunk 3]
```diff
@@ -44,6 +66,33 @@ def __init__(
...
+            logger.warning("Provider type is not set")
+            return
+
+        if provider not in available_provider_parameters:
+            logger.warning(
+                f"Available parameters for provider {provider} are not setup."
```
Comment (onmete) [inline ols/src/llms/providers/provider.py:81]: > f"Available parameters for provider {provider} are not defined. Parameters validation is disabled."

Or something like that to tell the user what is actually the outcome of that warning.

File: ols/src/llms/providers/provider.py (hunk 2 above)
Comment (tisnik) [reply ols/src/llms/providers/provider.py:75]: yeah was about to say that ;)

File: ols/src/llms/providers/provider.py (hunk 3 above)
Comment (tisnik) [reply ols/src/llms/providers/provider.py:81]: ok that makes sense, thank you @onmete

File: ols/src/llms/providers/provider.py (hunk 1 above)
Comment (tisnik) [reply ols/src/llms/providers/provider.py:18]: yup, dataclass is nice fit there

File: ols/src/llms/providers/provider.py (hunk 3 above)
Comment (tisnik) [reply ols/src/llms/providers/provider.py]: updated

File: ols/src/llms/providers/provider.py (hunk 1 above)
Comment (tisnik) [reply ols/src/llms/providers/provider.py:21]: implemented as dataclass now

File: ols/src/llms/providers/provider.py [hunk 4]
```diff
@@ -2,16 +2,92 @@
...
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass(frozen=True)
+class ProviderParameter:
+    """Specification of one provider parameter."""
+
+    _name: str
+    _type: type
```
Comment (onmete) [inline ols/src/llms/providers/provider.py:27]: Why are these underscored?

File: ols/src/llms/providers/provider.py [hunk 5]
```diff
@@ -43,7 +119,50 @@ def __init__(
...
+
+        # retrieve all available parameters for provider
+        available_parameters = available_provider_parameters[provider]
+
+        # only supported parameters will be returned
+        filtered_params: dict[str, Any] = {}
+
+        for parameter_name, parameter_value in params.items():
+            parameter_type = type(parameter_value)
+            if parameter_value is None:
```
Comment (onmete) [inline ols/src/llms/providers/provider.py:154]: This means that `"unknown_parameter": None` will go through the validation. Is this intended behavior?
Shoudn't it be inside the `if parameter not in available_parameters:` condition?

Comment (tisnik) [reply ols/src/llms/providers/provider.py:154]: ah actually it will be better, yes

File: ols/src/llms/providers/provider.py (hunk 4 above)
Comment (tisnik) [reply ols/src/llms/providers/provider.py:27]: `type` is sort of keyword

File: ols/src/llms/providers/provider.py (hunk 5 above)
Comment (tisnik) [reply ols/src/llms/providers/provider.py]: updated + added some unit tests

**Pull request merged**